contracts/UmaConditionalTokensBinaryAdapter.sol

```diff
@@ -525,89 +366,157 @@ contract UmaConditionalTokensBinaryAdapter is Auth, ReentrancyGuard {
     /// @notice Allows an authorized user to unpause market resolution in an emergency
     /// @param questionID - The unique questionID of the question
     function unPauseQuestion(bytes32 questionID) external auth {
-        require(isQuestionInitialized(questionID), "Adapter::unPauseQuestion: questionID is not initialized");
+        require(isQuestionInitialized(questionID), AdapterErrors.NotInitialized);
         QuestionData storage questionData = questions[questionID];
         questionData.paused = false;
         emit QuestionUnpaused(questionID);
     }
 
-    /// @notice Allows an authorized user to update the UMA Finder address
-    /// @param newFinderAddress - The new finder address
-    function setFinderAddress(address newFinderAddress) external auth {
-        emit NewFinderAddress(umaFinder, newFinderAddress);
-        umaFinder = newFinderAddress;
-    }
+    /*///////////////////////////////////////////////////////////////////
+                            INTERNAL FUNCTIONS 
+    //////////////////////////////////////////////////////////////////*/
 
-    /*
-    ////////////////////////////////////////////////////////////////////
-                            UTILITY FUNCTIONS 
-    ////////////////////////////////////////////////////////////////////
-    */
-
-    /// @notice Utility function that atomically prepares a question on the Conditional Tokens contract
-    ///         and initializes it on the Adapter
-    /// @dev Prepares the condition using the Adapter as the oracle and a fixed outcomeSlotCount
-    /// @param questionID               - The unique questionID of the question
-    /// @param ancillaryData            - Data used to resolve a question
-    /// @param resolutionTime           - Timestamp after which the Adapter can resolve a question
-    /// @param rewardToken              - ERC20 token address used for payment of rewards and fees
-    /// @param reward                   - Reward offered to a successful proposer
-    /// @param proposalBond             - Bond required to be posted by a price proposer and disputer
-    /// @param earlyResolutionEnabled   - Determines whether a question can be resolved early
-    function prepareAndInitialize(
+    function _saveQuestion(
         bytes32 questionID,
         bytes memory ancillaryData,
-        uint256 resolutionTime,
+        uint256 requestTimestamp,
         address rewardToken,
         uint256 reward,
-        uint256 proposalBond,
-        bool earlyResolutionEnabled
-    ) public {
-        conditionalTokenContract.prepareCondition(address(this), questionID, 2);
-        initializeQuestion(
-            questionID,
-            ancillaryData,
-            resolutionTime,
-            rewardToken,
-            reward,
-            proposalBond,
-            earlyResolutionEnabled
-        );
+        uint256 proposalBond
+    ) internal {
+        questions[questionID] = QuestionData({
```

Comment:
we use `questionID` here as the key, do you see any issues considering this could not be unique (ie same questionID different collateral)?

Reply:
if it doesn't, we can add collateral to the question data field

Reply:
is the data of `{rewardToken,reward,proposalBond}` present in the question data on IPFS?

Reply:
and by collateral, do you guys mean the rewardToken or collateral used on the CTF?

Reply:
I'm assuming collateral used on the CTF and rewardToken, reward and proposal bond are not stored on IPFS atm

Reply:
correct, with all this said. I don't think it's dangerous to assume uniqueness for questionID though there are some weird griefing vectors (ie front run in mem pool, deploying a market parameterized differently). If we agree we don't want to worry about this class of "attack" vectors I think it is fine as is.

Reply:
re: collateral, I don't think that the adapter needs to know anything about collateral used by the CTF (as questionId and conditionId are not related to collateral).
re: questionId, Any downside to having questionId simply be the keccak hash of ancillary data?

Reply:
or perhaps that would allow a griefing vector of front-running with a zero reward. could let questionId be the hash of ancillaryData together with reward and bond information